### PR TITLE
Adding a readme file to explain it all.

### PR DESCRIPTION
**Pull request by trothe**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+## Installation
+via Bower:
+
+    $ bower install git@github.com:travel-intelligence/ti-auth.js --save
+
+or via NPM:
+
+    $ npm install https://github.com/travel-intelligence/ti-auth.js --save-dev
+
+Alternatively you can manually download `dist/ti-auth.js` or `dist/ti-auth.min.js` and add it to your project.
```
Comment (jkintscher) [inline README.md]: I’d prefer not mentioning this given that we lose all versioning information. Especially since `master` can be in an unpublished state (even with versions/tags pointing to commits on `master` the latest commit should not be considered a stable release, as discussed).

If you really want to include this section, please update it to just provide the link to GitHub’s Releases page, here: https://github.com/travel-intelligence/ti-auth.js/releases

Comment (jkintscher) [inline README.md]: This got me thinking: Maybe we should include each release’s version number in a comment, along with copyright/licensing information, in a comment of the dist files?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+## Quick Start
+
+For ease of use the library exposes one global JavaScript object called **TiAuth**.<br>
```
Comment (jkintscher) [inline README.md]: Remove the trailing `<br>` as it’s not needed in Markdown.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+After including the library in the project page only 2 simple steps are needed to use `ti-auth.js` for authentication:
```
Comment (jkintscher) [inline README.md]: > in the project page

should probably be `in your application’s index.html file`

File: README.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+    ```javascript
+    TiAuth.API_URL = 'https://api.travel-intelligence.com';
+    TiAuth.DASHBOARD_URL = 'https://portal.travel-intelligence.com';
+    ```
+
+2. **Use the `initialize` method** - to bootstrap your application:
+
+    ```javascript
+    TiAuth.initialize(function(token, context) {
```
Comment (jkintscher) [inline README.md]: `context` sounds a little misleading here. Let’s just call the parameter `user` for more clarity and not to confuse it with a variable used for scope.

File: README.md [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+    $ bower install git@github.com:travel-intelligence/ti-auth.js --save
+
+or via NPM:
+
+    $ npm install https://github.com/travel-intelligence/ti-auth.js --save-dev
+
+## Quick Start
+
+For ease of use the library exposes one global JavaScript object called **TiAuth**.
```
Comment (jkintscher) [inline README.md]: Let’s use inline code formatting rather than bolding for `TiAuth` here.

File: README.md [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+After including the library in the in your application’s index.html file only 2 simple steps are needed to use `ti-auth.js` for authentication:
+
+1. **Configuration** - depending on the target environment (production/staging/local etc.) the Travel Intelligence API and the
```
Comment (jkintscher) [inline README.md]: This sentence sounds confusing. The interface does not reflect anything about the environment so I’d rather not even mention it here. Also because we actually want to encourage cross-environment usage, *i.e.* use the staging API/Dashboard when working in  development.

How about simplifying this to:

---

1. **Configuration** - You need to specify the URLs of the Travel Intelligence API and the Portal you would like to use:

  ```javascript
  TiAuth.API_URL = 'https://api.travel-intelligence.com';
  TiAuth.DASHBOARD_URL = 'https://portal.travel-intelligence.com';
  ```

  *Note:* If you fail to provide these two URLs before calling `TiAuth.initialize()` an exception is thrown.

File: README.md (hunk 5 above)
Comment (jkintscher) [inline README.md]: Let’s also add `[…] object called TiAuth that you interact with.` ?

File: README.md [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+    $ make
+
+
+---
+## Author
+Joschka Kintscher
+
+## MIT License
+Copyright (c) 2011 by Brian Noguchi
```
Comment (jkintscher) [inline README.md]: Who is this? And it’s 2016.

File: README.md [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    $ make test
+
+### Build
+
+    $ make
+
+
+---
+## Author
+Joschka Kintscher
```
Comment (jkintscher) [inline README.md]: Let’s link to https://github.com/jkintscher here:

```markdown
[Joschka Kintscher](https://github.com/jkintscher)
```

File: README.md (hunk 6 above)
Comment (jkintscher) [inline README.md]: Oh, just another note here that it could be confusing to mention the `DASHBOARD_URL` as `https://portal.travel-intelligence.com`. It’s currently still `https://travel-intelligence.com`, right?

File: README.md [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+1. **Configuration** - depending on the target environment (production/staging/local etc.) the Travel Intelligence API and the
+Portal URL need to be set. This can be done anywhere **before** the first call of the `TiAuth.initialize()`:
+
+    ```javascript
+    TiAuth.API_URL = 'https://api.travel-intelligence.com';
+    TiAuth.DASHBOARD_URL = 'https://portal.travel-intelligence.com';
+    ```
+
+2. **Use the `initialize` method** - to bootstrap your application:
```
Comment (jkintscher) [inline README.md]: Let’s make this into

2. **Authentication**: Verify the current user is signed in and start your application:

  ```js
  TiAuth.initialize(function(token, user) {
    // bootstrap and start your application here
  });
  ```

File: README.md [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+    TiAuth.initialize(function(token, user) {
+      //bootstrap your application here
+    });
+    ```
+
+  `token` is the authentication token needed to call Travel Intelligence services
+
+  `user` contains the user with its preferences
+
+## Logout
```
Comment (jkintscher) [inline README.md]: Let’s add a heading before this called

```markdown
## API
```

Where we’ll cover all of the public interfaces of `TiAuth`.

File: README.md [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+
+`ti-auth.js` provides the ability to invalidate the current user session. It will automatically redirect the user to the login page.
```
Comment (jkintscher) [inline README.md]: Let’s rephrase this to:

To sign out the current user and take them back to the login page of the TI Dashboard call

```js
TiAuth.signout();
```

File: README.md [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+
+```javascript
+TiAuth.logout();
+```
+
```
Comment (jkintscher) [inline README.md]: We should also cover the other public APIs of `TiAuth` here. So let’s add

---

### reauthorize

This allows you to make sure the current user is still authenticated. This is especially helpful when a request your module makes is rejected by the API with a `401` or `403` status code, indicating that the user’s token is no longer valid.

Example:

```js
TiAuth.reauthorize();
```

*Note that this will redirect them to the Dashboard and then back to your module if they are still logged in, so make sure to persist any state you want to return them to before calling this.*

### initialize

This is the main interface that allows you to ensure the user is authenticated before booting your application. It also gets you access to the user’s token and other information as provided by the API.

Example:

```js
TiAuth.initialize(function(token, user) {
  // bootstrap and start your application here
});
```

File: README.md [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+## Framework Integration
+`ti-auth.js` is build to be independent from any specific framework. This enables either a stand-alone use or the integration with common JavaScript frameworks.
```
Comment (jkintscher) [inline README.md]: How about

`ti-auth.js` does not have any dependencies and works completely independent from frameworks and libraries. At the same time it allows for seamless integration with any framework you may choose for your module. In this section you can find some example implementations for a few popular frameworks.

File: README.md [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+
+### Ember
+
+The initialization can be done in ember by creating a custom initializer.
```
Comment (jkintscher) [inline README.md]: Capitalize Ember.

File: README.md [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+
+1. Import `ti-auth.js` in `ember-cli-build.js`:
+ 
+  ```javascript
+  app.import('bower_components/ti-auth.js/dist/ti-auth.min.js');
```
Comment (jkintscher) [inline README.md]: Let’s change the path to a more generic one like `'path/to/your/ti-auth.js'` to avoid confusion, *i.e.* when Bower is configured with a different directory or if they use NPM.

File: README.md [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+
+      TiAuth.initialize(function(token, user) {
+        application.set('token', token);
+        application.advanceReadiness();
+      });
+    }
+  };
+
+  export
+  default TiAuthInitializer;
```
Comment (jkintscher) [inline README.md]: This should be one line.

File: README.md [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+1. Import `ti-auth.js` in `ember-cli-build.js`:
+ 
+  ```javascript
+  app.import('bower_components/ti-auth.js/dist/ti-auth.min.js');
+  ```
+2. Create `app/initializers/ti-auth.js` with the following content:
+  ```javascript
+  import Ember from 'ember';
+
+  var TiAuthInitializer = {
```
Comment (jkintscher) [inline README.md]: Actually, you can just make this

```js
export default {
```

and get rid of the `TiAuthInitializer` variable altogether.

File: README.md [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+    name: 'waitForTiAuth',
+    initialize: function(registry, application) {
```
Comment (jkintscher) [inline README.md]: As of Ember 2.0 the `initialize` function only expects one argument or else [will throw a deprecation warning](http://emberjs.com/deprecations/v2.x/#toc_initializer-arity), so you should change this to

```js
initialize: function(application) {
```

File: README.md [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+
+      application.deferReadiness();
+
+      TiAuth.API_URL = 'https://staging-api.travel-intelligence.com';
+      TiAuth.DASHBOARD_URL = 'http://travel-intelligence.dev';
```
Comment (jkintscher) [inline README.md]: Probably weird to mentioned a `.dev` URL here. Let’s just use https://travel-intelligence.com like above.

File: README.md [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+  ```javascript
+  import Ember from 'ember';
+
+  var TiAuthInitializer = {
+    name: 'waitForTiAuth',
+    initialize: function(registry, application) {
+
+      application.deferReadiness();
+
+      TiAuth.API_URL = 'https://staging-api.travel-intelligence.com';
```
Comment (jkintscher) [inline README.md]: And this should probably just use https://api.travel-intelligence.com.

File: README.md [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+        application.set('token', token);
+        application.advanceReadiness();
+      });
+    }
+  };
+
+  export
+  default TiAuthInitializer;
+  ```
+3. Import the initializer in the header of `app/app.js`:
```
Comment (jkintscher) [inline README.md]: I don’t think this is necessary (or correct) at all, is it? As long as the instance initializer is properly named and saved in `app/instance-initializers/ti-auth.js` all should be loaded automatically. So as long as you adjust the file path you mentioned above under 2 this whole step could be removed.

Comment (jkintscher) [inline README.md]: You could actually also mention that the file can be created with Ember-CLI using the following command:

```sh
$ ember generate instance-initializer ti-auth
```

which would be even more conventional.

File: README.md [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+  .. and load it with the other modules:
+  ```javascript
+  loadInitializers(App, config.modulePrefix, TiAuthInitializer);
+  ```
+
+### Angular
+
+1. Remove the `ng-app` attribute from the page root element.
+
+2. Manually bootstrap the angular app in the call-back after initializing `ti-auth.js`.
```
Comment (jkintscher) [inline README.md]: It’s just `callback`

File: README.md [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+
+    $ npm install
+
+### Tests
+
+    $ make test
+
+### Build
+
+    $ make
```
Comment (jkintscher) [inline README.md]: Let’s add

---

or use

    $ make watch

to automatically watch for file changes and re-build the library.

### Distribute

To create a distributable file (that includes all modules and exposes the `TiAuth` object), run

    $ make dist

File: README.md [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+
+
+---
```
Comment (jkintscher) [inline README.md]: Let’s remove this horizontal line.

File: README.md (hunk 23 above)
Comment (jkintscher) [inline README.md]: This should still be added (the **Distribute** part).

File: README.md (hunk 12 above)
Comment (jkintscher) [inline README.md]: :green_heart: This is already addressed, just didn’t auto-close the comment. Nothing more to be done.

File: README.md (hunk 10 above)
Comment (jkintscher) [inline README.md]: :green_heart: The API section has also been added. All done here.

File: README.md (hunk 9 above)
Comment (jkintscher) [inline README.md]: :green_heart: This has also been updated.

File: README.md (hunk 8 above)
Comment (jkintscher) [inline README.md]: :green_heart: Also done.

File: README.md (hunk 7 above)
Comment (jkintscher) [inline README.md]: :green_heart: Also done.

File: README.md (hunk 6 above)
Comment (jkintscher) [inline README.md]: :green_heart: Also done.

**Comment by jkintscher** [comment]
Hey @trothe I made a few further edits (see commits) as I felt like that’s the faster option than async review on a document this large. If you disagree with any of them or want to make further changes, feel free to amend or edit. Otherwise, this has my +1 in its current state now. Feel free to merge it.

I’ll then release a new version of the library including the README along with the [other changes](https://github.com/travel-intelligence/ti-auth.js/pull/9) (and potentially the [Grants](https://trello.com/c/WQC9e8Rv/26-user-grants-and-block-access) implementation) afterwards.

**Comment by trothe** [comment]
Thanks @jkintscher for making the changes directly in the file. I lost a bit the overview :)  
Now it's perfect and I'll go ahead and merge it!
